### PR TITLE
Validate `BaseUri` keys to match the `VersionedUri`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

URIs in `"$ref"` must match the base URI by which it is referenced. This PR validates the URIs.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202598517577618/f) _(internal)_

## 🔍 What does this change?

- It isn't straight forward to implement this validation, because the versioned URI lives in `*TypeReference`s, but the key could be anywhere else. To work around this, this PR adds a module-scoped trait `ValidateUri` trait. This trait is implemented on the `*TypeReference`s are called when validating the wrapping structs.
- Implement `Object: Deserialize` by hand (using `ObjectRepr`) because the required trait bound
- Adjust tests to pass the validation
- Add a test to check the validation

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs [hunk 1]
```diff
@@ -26,6 +30,19 @@ impl DataTypeReference {
     }
 }
 
+impl ValidateUri for DataTypeReference {
+    fn validate_uri(&self, base_uri: &BaseUri) -> Result<(), ValidationError> {
+        ensure!(
+            base_uri == self.reference.base_uri(),
+            ValidationError::BaseUriMismatch {
+                base_uri: base_uri.clone(),
+                versioned_uri: self.reference.clone()
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs:39]: Didn't we change this to self.uri for PropertyTypeReference? I thought we had done for DataTypeReference too but I must have missed it. Can we drive-by?

File: packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs [hunk 2]
```diff
@@ -24,6 +24,12 @@ pub enum ValidationError {
     /// A schema has marked a property with a [`BaseUri`] as required but the [`BaseUri`] does not
     /// exist in the `properties`.
     MissingRequiredProperty(BaseUri),
+    /// Referring to a [`VersionUri`] from a [`BaseUri`] requires that the [`VersionUri::base_uri`]
+    /// matches the [`BaseUri`].
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs]: Perhaps something more like:
```suggestion
    /// When associating a property name with a reference to a Type, we expect the name to match
    /// the [`VersionedUri::base_uri`] inside the reference
```

Or do we validate the base/versioned URIs on things that aren't just property names?

File: packages/graph/hash_graph/lib/graph/src/types/schema/object.rs [hunk 3]
```diff
@@ -20,14 +21,17 @@ struct ObjectRepr<V> {
     required: Vec<BaseUri>,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-#[serde(try_from = "ObjectRepr<V>", rename_all = "camelCase")]
+pub trait ValidateUri {
+    fn validate_uri(&self, base_uri: &BaseUri) -> Result<(), ValidationError>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/types/schema/object.rs:28]: This line was removed here. Was that intentional?
`#[serde(try_from = "ObjectRepr<V>", rename_all = "camelCase")]`

File: packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs [hunk 4]
```diff
@@ -38,21 +44,31 @@ impl fmt::Display for ValidationError {
...
                 )
             }
+            Self::BaseUriMismatch {
+                base_uri,
+                versioned_uri,
+            } => {
+                write!(
+                    fmt,
+                    "the base URI ({base_uri}) of the referenced versioned URI ({versioned_uri}) \
+                     is not matching"
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs:61]: I don't think this is clear. I think we should have something more like:

"expected base URI `{base_uri}` differed from the base URI of `{versioned_uri}`"

There's probably a better way to phrase it but I think the current one is quite confusing.

File: packages/graph/hash_graph/lib/graph/src/types/schema/object.rs [hunk 5]
```diff
@@ -59,17 +63,24 @@ impl<V, const MIN: usize> Object<V, MIN> {
...
             }
+        );
+
+        for uri in self.required() {
+            ensure!(
+                self.properties().contains_key(uri),
+                ValidationError::MissingRequiredProperty(uri.clone())
+            );
+        }
+        for (base_uri, referenced) in self.properties() {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/types/schema/object.rs:80]: This is a `reference` rather than a `referenced` right?

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/schema/object.rs:80]: (As in the object _is_ the reference, not the thing being referenced)

**Review by Alfred-Mountfield — DISMISSED** [review]

File: packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs:39]: I now see that we actually don't seem to have it for `PropertyTypeReference` _or_ `DataTypeReference. I think you force pushed and removed the commits from #790 🙃

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs:39]: The commit links are broken in your previous messages:
https://github.com/hashintel/hash/pull/790#discussion_r919983087

File: packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs]: I don't know if we also want to check URIs in other locations. Currently, we don't.

File: packages/graph/hash_graph/lib/graph/src/types/schema/object.rs (hunk 5 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/schema/object.rs]: 5970aad2fe762a8cf33b64608ca473adeb46e75a

File: packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs (hunk 4 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs]: Could we add some delimiters around the URI still?

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/schema/mod.rs]: Applied in 50b816777f1da61569fdb47ed5dff349ceed9f7f

**Review by thehabbos007 — APPROVED** [review]

File: packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/types/schema/data_type.rs]: For clarification, #790 only renamed the methods, not the functions (which were on purpose). Off GH we discussed this and decided to also rename the fields but leaving `DataType::id`, `PropertyType::id`, and `EntityType::id`.
